server/src/main/java/org/elasticsearch/action/support/broadcast/node/BoundedDiagnosticRequestPermits.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.support.broadcast.node;
+
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.AdjustableSemaphore;
+
+/**
+ * This class guards the amount of bounded diagnostic requests a node can concurrently coordinate.
+ */
+public class BoundedDiagnosticRequestPermits {
+
+    public static final Setting<Integer> MAX_CONCURRENT_BOUNDED_DIAGNOSTIC_REQUESTS_PER_NODE = Setting.intSetting(
+        "cluster.max_concurrent_bounded_diagnostic_requests_per_node",
+        20,
```

Comment:
Out of curiosity, how did we arrive at 20 for the default value of this?

Reply:
That is kind of a guess, I was planning to check with you if you have a better estimation. My line of thought:

I assume that it's common to poll every second and that 10 seconds is a long time to wait for the request. Based on that to hit the limit of the calls if the limit was 10, you would need 10 requests to hit the same node and last more than 10 seconds. Then I doubled it just to be sure that this will only be triggered if there is real trouble. 

What do you think?

Reply:
The number I had in mind was more like 100.

The typical polling interval with things like Metricbeat is more like 10s, but it hits multiple endpoints in each cycle. Not sure exactly how many, but I expect it's ≤10. In addition there's going to be ad-hoc stats requests (e.g. running the support diagnostic tool). 20 feels a little tight to me. The point of this feature is as a last-ditch protection against runaway stats requests leading to an OOM, just in case we have a broken client that doesn't cancel on a timeout, or measures like #82685 etc aren't quite enough. In the cases I've investigated the node has usually managed to collect several hundred sets of responses before falling over.

Obviously the OOM point is a function of the heap size of the node, and the size of the responses, and therefore the type of request and the size and complexity of the cluster. Ideally we'd impose an ergonomic limit based on these factors, but in practice I don't think it's worth the effort: a flat 100 would fix all the cases I've seen whilst not limiting reasonable usage.